core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -1858,6 +1858,8 @@ public Expression implement(
         final Primitive primitive = Primitive.ofBoxOr(type0);
         if (primitive == null
             || type1 == BigDecimal.class
+            || isNumberOrString(type0) && isNumberOrString(type1)
+            && backupMethodName.equals("plus")
```

Comment:
```suggestion
            || (backupMethodName.equals("plus") 
              && isNumberOrString(type0) 
              && isNumberOrString(type1))
```
运算符检查前置，降低`isNumberOrString`的不必要开销